mods/parsoid.js

```diff
@@ -69,17 +70,67 @@ PSP.wrapContentReq = function(restbase, req, promise) {
         return res;
     }
 
-    if(!rp.revision || rbUtil.isTimeUUID(rp.revision) || /^latest$/.test(rp.revision)) {
+    if(!rp.revision && !req.query.sections || rbUtil.isTimeUUID(rp.revision)) {
         // we are dealing with the latest revision,
         // so no need to check it, as the latest
         // revision can never be supressed
         return promise.then(ensureCharsetInContentType);
     }
-    // bundle the promise together with a call to getRevisionInfo()
-    return P.all([promise, this.getRevisionInfo(restbase, req)]).then(function(resx) {
+    var reqs = {
+        content: promise,
+    };
+
+    if (rp.revision) {
+        // Bundle the promise together with a call to getRevisionInfo(). A
+        // failure in getRevisionInfo will abort the entire request.
+        reqs.revisionInfo = this.getRevisionInfo(restbase, req);
+    }
+
+    // If the format is HTML and sections were requested, also request section
+    // offsets
+    if (format === 'html' && req.query.sections) {
+        reqs.sectionOffsets = restbase.get({
+            uri: this.getBucketURI(rp, 'section.offsets', tid)
+        }).catch(function() {});
+    }
+
+    return P.props(reqs)
+    .then(function(responses) {
         // if we have reached this point,
         // it means access is not denied
-        return ensureCharsetInContentType(resx[0]);
+        if (format === 'html' && req.query.sections && responses.sectionOffsets) {
+            // Handle section requests
+            var sectionOffsets = responses.sectionOffsets.body;
+            var sections = req.query.sections.split(',').map(function(id) {
+                return id.trim();
+            });
+            var body = cheapBodyInnerHTML(responses.content.body.toString());
```

Comment:
Is it a fact that section offsets provided by Parsoid will always match our cheap body return? Probably worth to put a comment about it, if so.